File: java/Arrays/15_3sum.java
import java.util.*;

class threeSum{
    public static List<List<Integer>> threeSum(int[] nums){
        List<List<Integer>> result = new ArrayList<>();
        Arrays.sort(nums);
        int lengthh = nums.length;
        int left,right,sum;
        for ( int i = 0 ; i < lengthh-2 ; i ++){
            if (i>0 && nums[i]==nums[i-1]){
                continue;
            }
            left = i+1;
            right=lengthh-1;
            while(left<right){
                sum = nums[i] + nums[right] + nums[left];
                if (sum==0){
                    List<Integer> a = new ArrayList<>();
                    a.add(nums[i]);
                    a.add(nums[right]);
                    a.add(nums[left]);
                    result.add(a);
                    while(left<right){
                        if(nums[left]==nums[left+1]){
                            left++;
                        }
                        else{
                            left++;
                            break;
                        }
                    }
                    while(left<right){
                        if(nums[right]==nums[right-1]){
                            right--;
                        }
                        else{
                            right--;
                            break;
                        }
                    }
                }
                else if(sum<0){
                    left++;
                }
                else{
                    right--;
                }
            }
        }
        return result;
    }
    public static void main(String args[]){
        Scanner sc = new Scanner(System.in);
        int n;
        System.out.println("Enter the length of the array : ");
        n = sc.nextInt();
        int[] nums= new int[n];
        for(int i = 0 ;i < n ; i ++){
            nums[i]=sc.nextInt();
        }
        System.out.println("The output is : "+threeSum(nums));
    }
}

File: python/Arrays/15_3Sum.py
class _3Sum(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        length = len(nums)
        nums.sort()
        output=[]
        for i in range(length-2):
            if (i>=1 and nums[i]==nums[i-1]):
                continue
            l=i+1
            r=length-1
            while (l<r):
                a = nums[r]+nums[i]+nums[l]
                if (a==0):
                    output.append([nums[i],nums[l],nums[r]])
                    while (l<r and nums[l]==nums[l+1]):
                        l+=1
                    while (l<r and nums[r]==nums[r-1]):
                        r-=1
                    l+=1
                    r-=1
                elif (a<0):
                    l+=1
                else:
                    r-=1
        return output
                

def main():
    nums = eval(input("Enter this list : "))
    obj = _3Sum()
    print(obj.threeSum(nums))


if __name__ == "__main__":
    main()
        